pkg/process/process.go

```diff
@@ -85,38 +93,45 @@ func FindAllAncestorProcessIDsInSameCgroup(pid int) ([]int, error) {
 	}
 
 	// Update the process tree.
+	if err := t.populate(); err != nil {
+		return nil, fmt.Errorf("failed to populate process tree: %w", err)
+	}
+
+	// Tree is updated. Try to find the process again.
+	p, ok := t.tree[k]
+	if !ok {
+		return nil, fmt.Errorf("failed to find the process for PID %d", pid)
+	}
+	return findAncestorPIDsInSameCgroup(p)
+}
+
+func (t *Tree) populate() error {
```

Comment:
Until we prove that this is a perf problem, what do you think of just completely re-building the tree and returning it in this function? Then we don't have to worry about pruning as we rebuild from scratch anyway.

Since we profile every 10s only, I would also throttle rebuilding to at most every 10s (I would also only add this optimization if we see it necessary though).

Reply:
We can try. I'm happy to run this for a couple of hours on my local cluster and attach the profile here.

Reply:
![CleanShot 2023-02-16 at 21 31 34](https://user-images.githubusercontent.com/536449/219480279-1771b3cc-15d3-4c3f-a6b3-e57e9d30dcb6.png)

https://pprof.me/2ba7315/

It's taking the majority of the CPU time with these changes https://github.com/parca-dev/parca-agent/pull/1348/commits/20ac55d6c10d6d518b79f24cca88c83655ea3e10

I'll revert and introduce a lock for shared resource.

Reply:
Thank you for trying this! Shame, but I guess it makes sense.